packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1228 @@
+    super.didUpdateWidget(oldWidget);
+    if (widget.routeInformationProvider != oldWidget.routeInformationProvider ||
+      widget.backButtonDispatcher != oldWidget.backButtonDispatcher ||
+      widget.routeInformationParser != oldWidget.routeInformationParser ||
+      widget.routerDelegate != oldWidget.routerDelegate) {
+      _currentRouteInformationParserTransaction = Object();
+      _currentRouterDelegateTransaction = Object();
+    }
+    if (widget.routeInformationProvider != oldWidget.routeInformationProvider) {
+      oldWidget.routeInformationProvider?.removeListener(_handleRouteInformationProviderNotification);
+      widget.routeInformationProvider?.addListener(_handleRouteInformationProviderNotification);
+      if (oldWidget.routeInformationProvider?.value != widget.routeInformationProvider?.value) {
+        _handleRouteInformationProviderNotification();
+      }
+    }
+    if (widget.backButtonDispatcher != oldWidget.backButtonDispatcher) {
+      oldWidget?.backButtonDispatcher?.removeCallback(_handleBackButtonDispatcherNotification);
+      widget?.backButtonDispatcher?.addCallback(_handleBackButtonDispatcherNotification);
+    }
+    if (widget.routerDelegate != oldWidget.routerDelegate) {
+      oldWidget.routerDelegate.removeListener(_handleRouterDelegateNotification);
+      widget.routerDelegate.addListener(_handleRouterDelegateNotification);
+      _maybeNeedToReportRouteInformation();
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.routeInformationProvider?.removeListener(_handleRouteInformationProviderNotification);
+    widget?.backButtonDispatcher?.removeCallback(_handleBackButtonDispatcherNotification);
+    widget.routerDelegate.removeListener(_handleRouterDelegateNotification);
+    _currentRouteInformationParserTransaction = null;
+    _currentRouterDelegateTransaction = null;
+    super.dispose();
+  }
+
+  void _processInitialRoute() {
+    _currentRouteInformationParserTransaction = Object();
+    _currentRouterDelegateTransaction = Object();
+    widget.routeInformationParser.parseRouteInformation(widget.routeInformationProvider.value)
+      .then<T>(_verifyRouteInformationParserStillCurrent(_currentRouteInformationParserTransaction, widget))
+      .then<void>(widget.routerDelegate.setInitialRoutePath)
+      .then<void>(_verifyRouterDelegatePushStillCurrent(_currentRouterDelegateTransaction, widget))
+      .then<void>(_rebuild);
+  }
+
+  void _handleRouteInformationProviderNotification() {
+    _currentRouteInformationParserTransaction = Object();
+    _currentRouterDelegateTransaction = Object();
+    widget.routeInformationParser.parseRouteInformation(widget.routeInformationProvider.value)
+      .then<T>(_verifyRouteInformationParserStillCurrent(_currentRouteInformationParserTransaction, widget))
+      .then<void>(widget.routerDelegate.setNewRoutePath)
+      .then<void>(_verifyRouterDelegatePushStillCurrent(_currentRouterDelegateTransaction, widget))
+      .then<void>(_rebuild);
```

Comment:
Is this _rebuild necessary Wouldn't the routerDelegate notify its listeners in response to `setNewRoutePath` if a rebuild is required?

Reply:
I think we always want to rebuild at the end of the notification.
I feel it does not hurt the performance and it will not be error prompt as well.

When I was writing router delegate it make sense to notifying the router to rebuild when i am actively changing the app state. However, it seems weird if the router told router delegate something has changed, the route r delegate has to notify the router to rebuild after it update the app state.

This has tripped me a lot when i was rewriting the stock app and they are all tricky to figure out. I feel no harm to add the rebuild here.